Source/WebCore/animation/AnimationTimelinesController.cpp

```diff
@@ -310,53 +311,77 @@ void AnimationTimelinesController::maybeClearCachedCurrentTime()
+AnimationTimelinesController::NamedTimelineMapEntry* AnimationTimelinesController::determineTreeOrder(Vector<NamedTimelineMapEntry> elementAncestors, Element& matchElement) const
 {
-    auto it = m_nameToScrollTimelineMap.find(name);
-    if (it != m_nameToScrollTimelineMap.end()) {
-        auto& existingScrollTimeline = it->value;
-        existingScrollTimeline->setSource(&source);
-        existingScrollTimeline->setAxis(axis);
-    } else {
-        auto newScrollTimeline = ScrollTimeline::create(name, axis);
-        newScrollTimeline->setSource(&source);
-        m_nameToScrollTimelineMap.set(name, WTFMove(newScrollTimeline));
+    auto* element = &matchElement;
+    while (element) {
+        auto i = elementAncestors.findIf([&] (const NamedTimelineMapEntry& entry) {
+            return element == entry.element.ptr();
+        });
+        if (i != notFound)
+            return &elementAncestors.at(i);
+        element = element->parentElement();
     }
+    ASSERT_NOT_REACHED();
+    return nullptr;
 }
 
-void AnimationTimelinesController::unregisterNamedScrollTimeline(const AtomString& name)
+AnimationTimeline* AnimationTimelinesController::timeline(Vector<NamedTimelineMapEntry> entries, Element& matchElement) const
 {
-    m_nameToScrollTimelineMap.remove(name);
+    // https://drafts.csswg.org/scroll-animations-1/#timeline-scoping
+    // From the spec: A named scroll progress timeline or view progress timeline is referenceable by:
+    // 1. the name-declaring element itself
+    // 2. that element’s descendants
+    // If multiple elements have declared the same timeline name, the matching timeline is the one declared on the nearest element in tree order.
+    // In case of a name conflict on the same element, names declared later in the naming property (scroll-timeline-name, view-timeline-name) take
+    // precedence, and scroll progress timelines take precedence over view progress timelines.
+
+    Vector<NamedTimelineMapEntry> elementAncestors;
+    for (auto& entry : entries) {
+        if (entry.element.ptr() ==  &matchElement || matchElement.isDescendantOf(entry.element))
+            elementAncestors.append(entry);
+    }
+
+    if (elementAncestors.isEmpty())
+        return nullptr;
+    if (elementAncestors.size() == 1)
+        return elementAncestors.first().timeline.ptr();
+    return determineTreeOrder(elementAncestors, matchElement)->timeline.ptr();
 }
 
-ScrollTimeline* AnimationTimelinesController::scrollTimelineForName(const AtomString& name) const
+void AnimationTimelinesController::registerNamedScrollTimeline(const AtomString& name, Element& source, ScrollAxis axis)
 {
-    return m_nameToScrollTimelineMap.get(name);
+    auto newScrollTimeline = ScrollTimeline::create(name, axis);
+    newScrollTimeline->setSource(&source);
+
+    m_nameToTimelineMap.ensure(name, [] {
+        return Vector<NamedTimelineMapEntry> { };
+    }).iterator->value.append({ newScrollTimeline, Ref<Element> { source } });
```

Comment:
Is this correct? This will replace the timeline instance on all calls, which may break the relationship established with previously instantiated CSS Animations. The previous method would update the entry found in the map.

Reply:
Fixed.